### PR TITLE
improved xgboost wandb_callback

### PR DESCRIPTION
**Pull request by ayulockin**
https://wandb.atlassian.net/browse/WB-NNNN
https://wandb.atlassian.net/browse/CLI-NNNN
https://github.com/wandb/client/issues/NNNN

Description
-----------

What does the PR do?

Testing
-------

How was this PR tested?

Release Notes
-------------

Below, please enter user-facing release notes as one or more bullet points.
If your change is not user-visible, write `NO RELEASE NOTES` instead, with no bullet points.

------------- BEGIN RELEASE NOTES ------------------

------------- END RELEASE NOTES --------------------

### CODE REVIEW

File: wandb/integration/xgboost/xgboost.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    You can use `wandb_callback` to save the model as artifacts.
+
+    Arguments:
+        log_model: (boolean) if True save the model as artifacts.
+    """
+    def __init__(self,
+                 log_model=False):
+        if wandb.run is None:
+            raise wandb.Error("You must call wandb.init() before wamdb_callback()")
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:32]: Typo: `wamdb_callback`

File: wandb/integration/xgboost/xgboost.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                wandb.log({metric_name: log[-1]}, commit=False)
+        
+        wandb.log({})
+        
+        return False
+    
+    def _log_model_as_artifact(self, model):
+        model_name = f'{wandb.run.id}_model.json'
+        model_path = f"{wandb.run.dir}/{model_name}"
+        model_config = model.save_config()
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:58]: Doesn't look like `model_config` is used anywhere?

File: wandb/integration/xgboost/xgboost.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    You can use `wandb_callback` to save the model as artifacts.
+
+    Arguments:
+        log_model: (boolean) if True save the model as artifacts.
+    """
+    def __init__(self,
+                 log_model=False):
+        if wandb.run is None:
+            raise wandb.Error("You must call wandb.init() before wamdb_callback()")
+
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:33]: Can we log the model config in the init please? Something like:

```
config = model.get_config()
wandb.config.update(config)
```

File: wandb/integration/xgboost/xgboost.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    def __init__(self,
+                 log_model=False):
+        if wandb.run is None:
+            raise wandb.Error("You must call wandb.init() before wamdb_callback()")
+        
+        self.log_model = log_model
+
+    def after_training(self, model):
+        '''Run after training is finished.'''
+
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:38]: Can we log the best score and best iteration here?

Something like:

```
wandb.summary['best_X'] = model.best_score
wandb.summary['best_iteration'] = bst.best_iteration
```

File: wandb/integration/xgboost/xgboost.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import wandb
+import xgboost as xgb
+
+class wandb_callback(xgb.callback.TrainingCallback):
+    """`wandb_callback` automatically integrates XGBoost with wandb.
+
+    Example:
+        ```python
+        xg_reg.fit(X_train,
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:14]: need to define `xg_reg` right?

**Review by morganmcg1 — COMMENTED** [review]
Nice start, can you share an example dashboard?

File: wandb/integration/xgboost/xgboost.py (hunk 3 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:33]: There is no `get_config` attribute for a Booster object. `xgb.get_config` can be used to get global configuration. Docs [here](https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.get_config). But there's not much info here (see image). 

image

Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:33]: Able to log config before training starts. Do you think logging it before is equal to logging it after training?

image

**Comment by ayulockin** [comment]
Here's an example dashboard: https://wandb.ai/ayush-thakur/xgboost-housing-demo

File: wandb/integration/xgboost/xgboost.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        return False
+    
+    def _log_model_as_artifact(self, model):
+        model_name = f'{wandb.run.id}_model.json'
+        model_path = f"{wandb.run.dir}/{model_name}"
+        model.save_model(str(model_path))
+        
+        model_artifact = wandb.Artifact(name=model_name, type='model')
+        model_artifact.add_file(model_path)
+        wandb.run.log_artifact(model_artifact)
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:86]: can just call `wandb.log_artifact` here, lets minimize the use of `wandb.run....` unless its necessary

File: wandb/integration/xgboost/xgboost.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+        xg_reg = xgb.XGBRegressor(**bst_params)
+        xg_reg.fit(X_train, 
+            y_train, 
+            eval_set=[(X_test, y_test)],
+            callbacks=[wandb_callback()])
+        )
+        ```
+
+    `wandb_callback` will automatically log evaluation data from any metrics collected by XGBoost, like rmse.
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:33]: Lets re-style this a little:

```
 Passing `wandb_callback` to XGBoost will:
- log the booster model configuration to Weights & Biases
- log evaluation metrics collected by XGBoost, such as rmse, accuracy etc to Weights & Biases
- log training metric collected by XGBoost (if you provide training data to eval_set)
- save and upload your trained model to to Weights & Biases Artifacts (when `log_model = True`)
```

File: wandb/integration/xgboost/xgboost.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        ```
+
+    `wandb_callback` will automatically log evaluation data from any metrics collected by XGBoost, like rmse. 
+    It can also log training data if you provide training data to eval_set. This callback automatically
+    saves the configuration. 
+
+    You can use `wandb_callback` to save the model as artifacts.
+
+    Arguments:
+        log_model: (boolean) if True save the model as artifacts.
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:40]: `log_model: (boolean) if True save and upload the model to Weights & Biases Artifacts.`

File: wandb/integration/xgboost/xgboost.py (hunk 3 above)
Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py:46]: I think so yep

**Review by morganmcg1 — COMMENTED** [review]
A few small nits, but overall looking really good now!

File: wandb/integration/xgboost/xgboost.py (hunk 4 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:69]: This one took a while but it's possible. Actually `model.best_score` throws an error (maybe in older versions it work) inside the callback. 

The current solution is: 

```
if model.attr('best_score') is not None:
    wandb.log({'best_score': float(cast(str, model.attr('best_score'))), 
                      'best_iteration': int(cast(str, model.attr('best_iteration')))})      
```

The question now is - should I use summary (hard to discover by new users) or log (will be logged at the very last step)?

File: wandb/integration/xgboost/xgboost.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            self._log_feature_importance(model)
+            
+        # Log the best score and best iteration
+        if model.attr('best_score') is not None:
+            wandb.log({'best_score': float(cast(str, model.attr('best_score'))), 
+                       'best_iteration': int(cast(str, model.attr('best_iteration')))})            
+        
+        return model
+
+    def after_iteration(self, model, epoch, evals_log):
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:87]: Just a thought, if we really wanted to add some spice, we could add define_metric here too so they get extra metrics. So once we know what metrics are going to be logged you can call:

```
maximise_metrics = ['accuracy', 'f1_score', ...]
minimize_metrics = ['loss', 'rmse', ....]

for m in metrics:
    if m in minimize_metrics:
        wandb.define_metric(m, summary="min")`
   elif m in maximize_metrics:
        wandb.define_metric(m, summary="max")`
```

If the metrics are available in the model config this could also be called in the `init` function

A chunk of the above code would actually be great to keep in the client so we have a persistent list of metrics and whether they should be minimized or maximised

File: wandb/integration/xgboost/xgboost.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        
+        model_artifact = wandb.Artifact(name=model_name, type='model')
+        model_artifact.add_file(model_path)
+        wandb.log_artifact(model_artifact)
+        
+    def _log_feature_importance(self, model):
+        fi = model.get_score(importance_type=self.importance_type)
+        fi_data = [[k, fi[k]] for k in fi]
+        table = wandb.Table(data=fi_data, columns = ["Feature", "Importance"])
+        wandb.log({"Feature Importance" : wandb.plot.bar(table, "Feature",
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:112]: Lets log this as `Model Performance/Feature Importance` or similar so it gets its own section name in the UI (I really don't like the "Custom Charts" section title default as they're only "custom" to us and I think its confusing for the user

**Review by morganmcg1 — COMMENTED** [review]
Just a thought, this might break existing compatibility right as the current integration doesn’t require an init? 

Maybe we need to keep the existing code and add a warning that it will be deprecated and they should use the new callback

Maybe call the new callback `WandbCallback` to differentiate from `wandb_callback`

File: wandb/integration/xgboost/xgboost.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    - log feature importance plot when `log_feature_importance=True`.
+    
+    Arguments:
+        log_model: (boolean) if True save and upload the model to Weights & Biases Artifacts
+        log_feature_importance: (boolean) if True log a feature importance bar plot
+        importance_type: (str) one of {weight, gain, cover, total_gain, total_cover} for tree model. weight for linear model. 
+    """
+    def __init__(self,
+                 log_model=False,
+                 log_feature_importance=False,
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:50]: yeah it's not much of an overhead.

File: wandb/integration/xgboost/xgboost.py (hunk 4 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:71]: `model.best_score` throws an error even though the attribute exists. 

image

The code snippet below is from the official [repo](https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/callback.py#L155) and may highlight the issue. That's why I used the same way even though it's not very clean. 

image

File: wandb/integration/xgboost/xgboost.py (hunk 9 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:87]: This is a neat feature that we can add, but in this context where after iteration some metrics are to be logged, there is no sense of custom step. Correct me if I am missing on something here.

Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:87]: This is a great idea. Let's spice it up then.

File: wandb/integration/xgboost/xgboost.py (hunk 10 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py]: This is creating another issue. 

> Hey @support_team, when I log a bar chart like this, wandb.log({"Model Performance/Feature Importance" : wandb.plot.bar(table, "Feature", "Importance", title="Feature Importance")}) it's creating a custom chart under a panel section named "Model Performance". However, it's also creating Tables (of the data logged) under the same section. I would want this to go under a panel section named Tables. What can be done? Example run page: https://wandb.ai/ayush-thakur/xgboost-housing/runs/1ivxlgp7?workspace=user-ayush-thakur

**Comment by ayulockin** [comment]
@morganmcg1 

> Maybe we need to keep the existing code and add a warning that it will be deprecated and they should use the new callback

Added UserWarning that it will be deprecated. One user warning is coming from our side, the other one is XGBoost's.

image

> Maybe call the new callback WandbCallback to differentiate from wandb_callback

I have named the new callback XGBoostCallback for now. If you want to use WandbCallback for consistency I will put that.

File: wandb/integration/xgboost/xgboost.py (hunk 9 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:106]: How about we do it in the immediate version?

**Comment by ayulockin** [comment]
Also a colab to compare old and new callbacks: https://colab.research.google.com/drive/1dvAQmqkLP51IP8AXaqzDdbdYh7ruOQ5Z?usp=sharing

File: wandb/integration/xgboost/xgboost.py (hunk 4 above)
Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py:90]: well thats disappointing! original solution looks good then. 

Is it worth adding `model.best_ntree_limit` too since it looks like its available?

File: wandb/integration/xgboost/xgboost.py (hunk 9 above)
Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py:106]: sorry what do you mean by "immediate version"?

File: wandb/integration/xgboost/xgboost.py (hunk 10 above)
Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py]: ok maybe its cleaner to keep the original version which puts the chart and table in separate sections.

Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py]: ok lets keep the original default where the chart and the table are split into different sections

**Comment by morganmcg1** [comment]
Lets keep the name as `WandbCallback` as it aligns with all of our other integrations callbacks

File: wandb/integration/xgboost/xgboost.py (hunk 9 above)
Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py:106]: sorry what do you mean by "immediate version"? Lets manually create the min/max lists in this integration code for now so we can already ship this, and in parallel we can chat with the client team about how to include it in the client itself in the coming weeks

File: wandb/integration/xgboost/xgboost.py (hunk 4 above)
Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py:90]: lets go with your original version so

Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:90]: Added `best_ntree_limit`

Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:90]: Also note that adding scalar values as wandb.summary will not create a chart automatically. So need to use `wandb.log`.

Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:90]: I don't think `model.best_ntree_limit` is available off the shelf.

File: wandb/integration/xgboost/xgboost.py (hunk 9 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:113]: Added this. Have a check. Note that the metric is available only in `after_iteration` for the first epoch.

File: wandb/integration/xgboost/xgboost.py (hunk 4 above)
Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py:97]: > So need to use wandb.log
 sounds good

Lets exclude `best_ntree_limit` so

File: wandb/integration/xgboost/xgboost.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+                 log_feature_importance=True,
+                 importance_type='gain'):
+        
+        if wandb.run is None:
+            raise wandb.Error("You must call wandb.init() before WandbCallback()")
+        
+        self.log_model = log_model
+        self.log_feature_importance = log_feature_importance
+        self.importance_type = importance_type
+        self.define_metrics = False
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:84]: lets turn `define_metrics` on by default too, it also needs to be added as a parameter to `init`

File: wandb/integration/xgboost/xgboost.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        return model
+
+    def after_iteration(self, model, epoch, evals_log):
+        '''Run after each iteration. Return True when training should stop.'''
+  
+        # Log metrics
+        for data, metric in evals_log.items():
+            for metric_name, log in metric.items():
+                if not self.define_metrics:
+                    self._define_metrics(data, metric_name)
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:120]: I think this can be broken out to only be called during the first pass of `after_iteration`. 

```
def __init__(....):
    ....
    self.step_count = 0

.....

def after_iteration(...)
    ...
    for data, metric in evals_log.items():
        for metric_name, log in metric.items():
            if not self.define_metrics and self.step_count == 0:
                self._define_metrics(data, metric_name)

           wandb.log({f"{data}-{metric_name}": log[-1]}, commit=False

    self.step_count+=1

    ...

```

Also I think it should be `if self.define_metrics ...`, and not `if not self.define_metrics ...` right? In general its good to avoid using `if not ...` if there is an easier way of doing things, just for human interpretability

File: wandb/integration/xgboost/xgboost.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+        
+    def _log_feature_importance(self, model):
+        fi = model.get_score(importance_type=self.importance_type)
+        fi_data = [[k, fi[k]] for k in fi]
+        table = wandb.Table(data=fi_data, columns = ["Feature", "Importance"])
+        wandb.log({"Feature Importance" : wandb.plot.bar(table, "Feature",
+                                                         "Importance", title="Feature Importance")})
+        
+    def _define_metrics(self, data, metric_name):
+        if metric_name in MINIMIZE_METRICS:
```
Comment (morganmcg1) [inline wandb/integration/xgboost/xgboost.py:148]: Lets wrap metric name in lowercase to reduce the changes that we miss matching a metric.

I think we can also make a special case for any metric that contains "loss" too I think, e.g. 

```
if "loss" in lower(metric_name):
    wandb.define_metric(f"{data}-{metric_name}", summary="min")  
elif metric_name in MINIMIZE_METRICS:
    ...
elif metric_name in MAXIMIZE_METRICS:
   ...
```

File: wandb/integration/xgboost/xgboost.py (hunk 13 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:120]: The current solution is like,

By default `self.define_metrics` is True. In the `after_iteration` once the `_define_metrics` method is called, self.define_metric is turned False. I don't think we need to maintain a `self.step_count`.

Comment (morganmcg1) [reply wandb/integration/xgboost/xgboost.py]: oh nice, sorry I missed that!
